app/models/sftp/cli.rb

```diff
@@ -0,0 +1,266 @@
+    end
+
+    # Creates a temporary file with restricted permissions (0600) containing the password.
+    # This file is used by sshpass to avoid exposing the password in process lists or command history.
+    def ensure_password_file
+      return if @password_file && File.exist?(@password_file.path)
+
+      @password_file = Tempfile.new('sftp_password')
+      @password_file.write(@options[:password] || '')
+      @password_file.close
+      File.chmod(0o600, @password_file.path)
+    end
+
+    # Escapes special characters in file paths for safe use in SFTP batch commands.
+    # Note: This is for SFTP command escaping, not shell escaping.
+    def escape_path(path)
+      Shellwords.escape(path.to_s)
+    end
+
+    # Proxy for directory operations, providing Net::SFTP-compatible API.
+    class DirProxy
+      def initialize(connection)
+        @connection = connection
+      end
+
+      # Iterates over all entries in a directory, yielding RemoteFile objects.
+      def foreach(directory)
+        escaped_dir = @connection.send(:escape_path, directory)
+        output = @connection.send(:execute_sftp_with_output, "cd #{escaped_dir}\nls -la")
+        parse_ls_long_output(output).each { |entry| yield entry }
+      end
+
+      # Lists files in the given directory matching the glob pattern.
+      # Returns an array of RemoteFile objects with name attribute.
+      def glob(directory, pattern)
+        escaped_dir = @connection.send(:escape_path, directory)
+        output = @connection.send(:execute_sftp_with_output, "cd #{escaped_dir}\nls -1")
+        parse_ls_output(output, pattern)
+      end
+
+      private
+
+      # Parses the output of 'ls -la' command, preserving the full line as longname.
+      def parse_ls_long_output(output)
+        output.each_line.filter_map do |line|
+          line = line.strip
+          next if line.empty? || line.start_with?('sftp>')
+
+          # Extract filename from the end of ls -l output
+          # Format: -rw-r--r-- 1 user group 1234 Jan 15 10:30 filename
+          match = line.match(/\s(\S+)$/)
+          next unless match
+
+          RemoteFile.new(name: match[1], longname: line)
+        end
+      end
+
+      # Parses the output of 'ls -1' command (one filename per line).
+      # Filters files by the glob pattern and creates RemoteFile objects.
+      def parse_ls_output(output, pattern)
+        pattern_regex = glob_to_regex(pattern)
+
+        output.each_line.filter_map do |line|
+          filename = line.strip
+          next if filename.empty? || filename.start_with?('sftp>')
+
+          RemoteFile.new(name: filename, longname: filename) if filename.match?(pattern_regex)
+        end
+      end
+
+      # Converts a glob pattern (e.g., "*.csv") to a regex for matching filenames.
+      def glob_to_regex(pattern)
+        regex_str = pattern.gsub('.', '\.').gsub('*', '.*').gsub('?', '.')
+        Regexp.new("^#{regex_str}$", Regexp::IGNORECASE)
```

Comment:
technically safest to escape regex special characters in pattern?

```suggestion
      escaped = Regexp.escape(pattern)
      regex_str = escaped.gsub('\*', '.*').gsub('\?', '.')
      Regexp.new("^#{regex_str}$", Regexp::IGNORECASE)
```